specification/servicefabric/resource-manager/Microsoft.ServiceFabric/stable/2020-03-01/application.json

```diff
@@ -1395,6 +1426,61 @@
       "description": "The amount of time to wait after completing an upgrade domain before applying health policies. It is first interpreted as a string representing an ISO 8601 duration. If that fails, then it is interpreted as a number representing the total number of milliseconds.",
       "default": "0"
     },
+    "ManagedIdentity": {
+      "description": "Describes the managed identities for an Azure resource.",
+      "properties": {
+        "principalId": {
+          "type": "string",
+          "readOnly": true,
+          "description": "The principal id of the managed identity. This property will only be provided for a system assigned identity."
+        },
+        "tenantId": {
+          "type": "string",
+          "readOnly": true,
+          "description": "The tenant id of the managed identity. This property will only be provided for a system assigned identity."
+        },
+        "type": {
+          "$ref": "#/definitions/ManagedIdentityType",
+          "description": "The type of managed identity for the resource."
+        },
+        "userAssignedIdentities": {
+          "$ref": "#/definitions/UserAssignedIdentityMap",
+          "description": "The list of user identities associated with the resource. The user identity dictionary key references will be ARM resource ids in the form:\n'/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ManagedIdentity/userAssignedIdentities/{identityName}'.\n"
+        }
+      }
+    },
+    "ManagedIdentityType": {
+      "type": "string",
+      "description": "The type of managed identity for the resource.",
+      "enum": [
+        "SystemAssigned",
+        "UserAssigned",
```

Comment:
Could you share more context? Why do we need `identity` and `managedIdentity` both? Looks like `identity` already supports UserAssigned.

Reply:
Hi @MyronFanQiu - here's our document on this:
https://docs.microsoft.com/en-us/azure/service-fabric/how-to-deploy-service-fabric-application-user-assigned-managed-identity#user-assigned-identity

Basically that section maps user assigned identities to "friendly" names that services in an Service Fabric application can reference to. Also adding @dragav for any further clarifications.

Reply:
@MyronFanQiu, the 'identity' field is a standard element of the schema of Azure resources - it denotes support for managed identities, and as you pointed out, supports user- and system-assigned identities.

These identities need to be translated into properties of services which comprise the SF application as an Azure resource. The motivation behind it is that a service (which is not an Azure resource) may only be assigned/consume an identity assigned to its parent app (the application). As Meng explained, the 'managedIdentities' field is just a mapping from the 'identities' element into references declared in the service manifests.

Reply:
well. based on the doc, as a customer, I need to input the same identity information twice, which is not convenient for me. But anyway, it's fine. Could you change the description of `identity` if it only support user identity?

Reply:
It is what it is. The standard ARM Identity property does not specify a way to create aliases for identities. We have to find a way to map identities from the ARM resource definition to late-bound names declared by the services. The principal ids corresponding to the entries in the 'identity' property won't be known until after deployment.

'identity' supports both system- and user-assigned identities.

Also, the structure of the 'managedIdentities' property looks like this:

        "managedIdentities": [
          {
            "name" : "[parameters('userAssignedIdentityFriendlyName')]",
            "principalId" : "[reference(variables('userAssignedIdentityResourceId'), variables('managedIdentityApiVersion')).principalId]"
          }
        ]

which is an array of <name, principalId> tuples; since the system-assigned identity won't have its principal id known at deployment time (it's being created during the deployment), it will be referenced directly from the service manifest by the alias 'SystemIdentity' (and so does not need to be listed in the 'managedIdentities' array).

Reply:
got it. make senses. Please fix the CI errors and I can merge this one.